scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java

```diff
@@ -255,28 +255,7 @@ public boolean resume() {
     }
 
     public boolean shutdown() {
-        if (status.isDown()) {
-            return false;
-        }
-
-        status = SchedulerStatus.SHUTTING_DOWN;
-        logger.info("Scheduler is shutting down, this may take time to finish every jobs!");
-        listener.schedulerStateUpdated(SchedulerEvent.SHUTTING_DOWN);
-
-        logger.info("Unpause all running and pending jobs!");
-        jobs.unpauseAll();
-
-        infrastructure.schedule(new Runnable() {
-            public void run() {
-                if (jobs.getRunningTasks().isEmpty()) {
-                    listener.schedulerStateUpdated(SchedulerEvent.SHUTDOWN);
-                } else {
-                    infrastructure.schedule(this, 5000);
-                }
-            }
-        }, 5000);
-
-        return true;
+        return freeze() && kill();
```

Comment:
I think this will not be enough, because you must wait until there are no currently running tasks

Reply:
part of the previous code was actually ok, except we should not unpauseAll I think, and we should call kill when there are no running tasks. We must see which statuses become obsolete (I think the status SHUTING_DOWN makes sense, but I don't know if we use the status SHUTDOWN or KILLED during the final step)

Reply:
Thx for this comment.. but not sure I completely understood. Freezing the scheduler already do the job: it makes the scheduler wait for running tasks to be terminated. Moreover we said that shutdown <=> freeze + kill.  For your second comment, as i can see for the other admin actions (freeze, pause), the status is updated before executing the procedure to freeze/pause.. In our case, we would set the scheduler status to SHUTDOWN before shuting down it.